src/macros.jl

```diff
@@ -0,0 +1,125 @@
+assigned at runtime from an `__init__()` function:
+```julia
+function __init__()
+    H5PREFIX.DYANMIC_CONSTANT = my_runtime_value()
+end
+```
+"""
+macro defconstants(prefix::Symbol, expr::Expr)
+    isexpr(expr, :block) || error("Expected block expression")
+    stmts = expr.args
+
+    symbols = Vector{Symbol}(undef, 0)
+    imports = Vector{Union{Symbol,Expr}}(undef, 0)
+    defbody = Vector{Expr}(undef, 0)
+    getbody = Vector{Expr}(undef, 0)
+    setbody = Vector{Expr}(undef, 0)
+
+    innermod = Symbol(:_, prefix)
+    for line in stmts
+        line isa LineNumberNode && continue
+
+        isruntime = isexpr(line, :(=)) ? false :
+                    isexpr(line, :(::)) ? true :
+                    error("Unexpected statement: ", line)
+
+        # Get the name and type pair
+        nametype = isruntime ? line : line.args[1]
+        isexpr(nametype, :(::)) || error("Expected `name::type`, got: ", nametype)
+        name = nametype.args[1]::Symbol
+        type = nametype.args[2]::Union{Symbol,Expr}
+        # Save type for later use
+        push!(imports, type)
+
+        if isruntime
+            value = esc(:(Ref{$type}()))
+        else
+            valexpr = :(convert($type, $(line.args[2])))
+            value = QuoteNode(Core.eval(__module__, valexpr))
+        end
+        fullname = esc(:($innermod.$name))
+        getexpr = isruntime ? :($(fullname)[]) : fullname
+
+        push!(symbols, name)
+        push!(defbody, :(const $(esc(name)) = $value))
+        push!(getbody, :(sym === $(QuoteNode(name)) && return $getexpr))
+        if isruntime
+            setexpr = quote
+                sym === $(QuoteNode(name)) && begin
+                    $fullname[] = value
+                    return value
+                end
+            end
+            append!(setbody, Base.remove_linenums!(setexpr).args)
+        end
+    end
+
+    # Build expressions to import all necessary types from the parent module.
+    push!(imports, :Ref) # explicitly add `Ref`
+    imports = Expr[Expr(:(.), :(.), :(.), sym) for sym in unique!(imports)]
+    # Push the imports into the definition body
+    pushfirst!(defbody, Expr(:import, imports...))
+
+    eprefix = esc(prefix)
+    einnermod = esc(innermod)
+    block = quote
+        baremodule $einnermod
+            struct $eprefix end
+            $(defbody...)
+        end
+        const $eprefix = $einnermod.$prefix()
+        function Base.propertynames(::$einnermod.$prefix)
+            return $((symbols...,))
+        end
+        function Base.getproperty(::$einnermod.$prefix, sym::Symbol)
+            $(getbody...)
+            error($(string(prefix) * " has no constant "), sym)
+        end
+    end
+    if !isempty(setbody)
+        setfn = quote
+            function Base.setproperty!(::$einnermod.$prefix, sym::Symbol, value)
+                $(setbody...)
+                error($(string(prefix) * "."), sym, " cannot be set")
```

Comment:
We should probably change the error message to be consistent with what is printed in Base. Either:

Assigning a variable to in an other module:
"ERROR: cannot assign variables in other modules"

Trying to redefine a constant:
"ERROR: cannot declare x constant; it already has a value"

Reply:
I like the second one because it's still technically correct.